File: projects/py/src/sonic_ws/packet_logger.py
import logging
import time


class PacketLogger:
    """Readable packet logging middleware for clients and server connections."""

    def __init__(self, logger=None, include_values=True):
        self.logger = logger or logging.getLogger("sonic_ws.packets").info
        self.include_values = include_values
        self._sends = {}
        self._receive_sizes = {}

    def onSend_pre(self, tag, values, *_):
        self._sends[tag] = values

    def onSend_post(self, tag, _data, size):
        self._emit("send", tag, self._sends.pop(tag, None), size + 1)

    def onReceive_pre(self, tag, _data, size):
        self._receive_sizes.setdefault(tag, []).append(size + 1)

    def onReceive_post(self, tag, values):
        queue = self._receive_sizes.get(tag, [])
        size = queue.pop(0) if queue else 0
        if not queue:
            self._receive_sizes.pop(tag, None)
        self._emit("receive", tag, values, size)

    def _emit(self, direction, tag, values, size):
        self.logger({
            "direction": direction,
            "tag": tag,
            "values": values if self.include_values else None,
            "bytes": size,
            "timestamp": time.time(),
        })

File: projects/ts/src/ws/debug/PacketLogger.ts
import type { ConnectionMiddleware } from "../PacketProcessor";

export type PacketLogEntry = {
    direction: "send" | "receive";
    tag: string;
    values: unknown;
    bytes: number;
    timestamp: number;
};

export type PacketLoggerOptions = {
    logger?: (entry: PacketLogEntry) => void;
    includeValues?: boolean;
};

/** Readable packet logging implemented through the public middleware hooks. */
export class PacketLogger implements ConnectionMiddleware {
    private readonly logger: (entry: PacketLogEntry) => void;
    private readonly includeValues: boolean;
    private readonly sends = new Map<string, unknown[]>();
    private readonly receiveSizes = new Map<string, number[]>();

    constructor(options: PacketLoggerOptions = {}) {
        this.includeValues = options.includeValues ?? true;
        this.logger = options.logger ?? (entry => {
            const arrow = entry.direction === "send" ? "→" : "←";
            console.debug(`${arrow} ${entry.tag}`, entry.values, `${entry.bytes} bytes`);
        });
    }

    onSend_pre = (tag: string, values: unknown[]): void => { this.sends.set(tag, values); };
    onSend_post = (tag: string, _data: Uint8Array, size: number): void => {
        this.emit("send", tag, this.sends.get(tag), size + 1);
        this.sends.delete(tag);
    };
    onReceive_pre = (tag: string, _data: Uint8Array, size: number): void => {
        const queue = this.receiveSizes.get(tag) ?? [];
        queue.push(size + 1);
        this.receiveSizes.set(tag, queue);
    };
    onReceive_post = (tag: string, values: unknown): void => {
        const queue = this.receiveSizes.get(tag);
        this.emit("receive", tag, values, queue?.shift() ?? 0);
        if (queue?.length === 0) this.receiveSizes.delete(tag);
    };
    private emit(direction: "send" | "receive", tag: string, values: unknown, bytes: number): void {
        this.logger({ direction, tag, values: this.includeValues ? values : undefined, bytes, timestamp: Date.now() });
    }
}
